apps/core/src/hooks/useFormatCoin.ts

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { Coin, JsonRpcProvider } from '@mysten/sui.js';
+import { useQuery, type UseQueryResult } from '@tanstack/react-query';
+import BigNumber from 'bignumber.js';
+import { useMemo } from 'react';
+
+import { formatAmount } from '../utils/formatAmount';
+
+type FormattedCoin = [
+    formattedBalance: string,
+    coinSymbol: string,
+    queryResult: UseQueryResult
+];
+
+export enum CoinFormat {
+    ROUNDED = 'ROUNDED',
+    FULL = 'FULL',
+}
+
+/**
+ * Formats a coin balance based on our standard coin display logic.
+ * If the balance is less than 1, it will be displayed in its full decimal form.
+ * For values greater than 1, it will be truncated to 3 decimal places.
+ */
+export function formatBalance(
+    balance: bigint | number | string,
+    decimals: number,
+    format: CoinFormat = CoinFormat.ROUNDED
+) {
+    const bn = new BigNumber(balance.toString()).shiftedBy(-1 * decimals);
+
+    if (format === CoinFormat.FULL) {
+        return bn.toFormat();
+    }
+
+    return formatAmount(bn);
+}
+
+export function useCoinDecimals(
+    rpcClient: JsonRpcProvider,
+    coinType?: string | null
+) {
+    const queryResult = useQuery(
+        ['denomination', coinType],
+        async () => {
+            if (!coinType) {
+                throw new Error(
+                    'Fetching coin denomination should be disabled when coin type is disabled.'
+                );
+            }
+
+            return rpcClient.getCoinMetadata(coinType);
+        },
+        {
+            // This is currently expected to fail for non-SUI tokens, so disable retries:
+            retry: false,
+            enabled: !!coinType,
+            // Never consider this data to be stale:
+            staleTime: Infinity,
+            // Keep this data in the cache for 24 hours.
+            // We allow this to be GC'd after a very long time to avoid unbounded cache growth.
+            cacheTime: 24 * 60 * 60 * 1000,
+        }
+    );
+
+    return [queryResult.data?.decimals || 0, queryResult] as const;
+}
+
+// TODO #1: This handles undefined values to make it easier to integrate with
+// the reset of the app as it is today, but it really shouldn't in a perfect world.
+// TODO #2: In another perfect world, we would retrieve RPC client providers across
```

Comment:
We could pretty easily add an `RPCProvider` in the core + define a `useRpc()` hook (which we can then extract out of the individual apps). Although we can also do that as a follow-up.

Reply:
I went ahead and did this in this changeset since getting CI to pass with the other approach was just as much work